paimon-core/src/main/java/org/apache/paimon/catalog/CachingCatalog.java

```diff
@@ -0,0 +1,194 @@
+                if (identifier.getDatabaseName().equals(name)) {
+                    tables.add(identifier);
+                }
+            }
+            tables.forEach(tableCache::invalidate);
+        }
+    }
+
+    @Override
+    public void dropTable(Identifier identifier, boolean ignoreIfNotExists)
+            throws TableNotExistException {
+        super.dropTable(identifier, ignoreIfNotExists);
+        invalidateTable(identifier);
+    }
+
+    @Override
+    public void renameTable(Identifier fromTable, Identifier toTable, boolean ignoreIfNotExists)
+            throws TableNotExistException, TableAlreadyExistException {
+        super.renameTable(fromTable, toTable, ignoreIfNotExists);
+        invalidateTable(fromTable);
+    }
+
+    @Override
+    public void alterTable(
+            Identifier identifier, List<SchemaChange> changes, boolean ignoreIfNotExists)
+            throws TableNotExistException, ColumnAlreadyExistException, ColumnNotExistException {
+        super.alterTable(identifier, changes, ignoreIfNotExists);
+        invalidateTable(identifier);
+    }
+
+    @Override
+    public Table getTable(Identifier identifier) throws TableNotExistException {
+        Table table = tableCache.getIfPresent(identifier);
+        if (table != null) {
+            return table;
+        }
+
+        if (isSpecifiedSystemTable(identifier)) {
+            String[] splits = tableAndSystemName(identifier);
+            String tableName = splits[0];
+            String type = splits[1];
+
+            Identifier originIdentifier =
+                    Identifier.create(identifier.getDatabaseName(), tableName);
+            Table originTable = tableCache.getIfPresent(originIdentifier);
+            if (originTable == null) {
+                originTable = wrapped.getTable(originIdentifier);
+                tableCache.put(originIdentifier, originTable);
+            }
+            table = SystemTableLoader.load(type, (FileStoreTable) originTable);
+            if (table == null) {
+                throw new TableNotExistException(identifier);
+            }
+            tableCache.put(identifier, table);
+            return table;
+        }
+
+        table = wrapped.getTable(identifier);
+        tableCache.put(identifier, table);
+        return table;
+    }
+
+    private class TableInvalidatingRemovalListener implements RemovalListener<Identifier, Table> {
+        @Override
+        public void onRemoval(
+                Identifier tableIdentifier, Table table, @NonNull RemovalCause cause) {
+            LOG.debug("Evicted {} from the table cache ({})", tableIdentifier, cause);
+            if (RemovalCause.EXPIRED.equals(cause)) {
+                if (!isSpecifiedSystemTable(tableIdentifier)) {
+                    tableCache.invalidateAll(allSystemTables(tableIdentifier));
+                }
+            }
+        }
+    }
+
+    private void invalidateTable(Identifier identifier) {
```

Comment:
We can expose the method. For example, Spark `TableCatalog` has an interface which should uses it
```scala
  /**
   * Invalidate cached table metadata for an {@link Identifier identifier}.
   * <p>
   * If the table is already loaded or cached, drop cached data. If the table does not exist or is
   * not cached, do nothing. Calling this method should not query remote services.
   *
   * @param ident a table identifier
   */
  default void invalidateTable(Identifier ident) {
  }
```

Reply:
Good point, fixed.